### DISCUSSION TITLE
React Router: how to get the redirect `Response` of a redirect call? when returned/thrown from loader?

### DISCUSSION THREAD

**Discussion by svenvandescheur** [remix-run/react-router#11715] [Proposals]
# Scenario
 
- Given a case where a loader attempts to fetch data for a Route
 - And some validation check fails causing a `redirect` with additional information passed to the the `init` parameter (`ResponseInit`)
 - When the new (redirected) Route is loaded
 - Then I need access to the `Response` to explain the reason for the redirect to the user.
 
 # Pseudo code
 
 ```js
// This one causes a redirect `Response` to be returned.
async function authenticatedPageLoader() {
  // Some auth check fails...
  return redirect("/login", {statusText: "Please log in!");
} 

// This one receives `Params` and a `Request` without any information about the reason why the redirect happened which is passed to the `Response`.
async function loginLoader({params, request}) {
  // Explain the user that they need to be logged in to visit "/some/authenticated/page"
} 

createBrowserRouter([
  {
    path: "/some/authenticated/page",
    element: <AuthenticatedPage/>
    loader: authenticatedPageLoader,
  },
  {
    path: "/login",
    element: <LoginPage/>
    loader: loginLoader, 
  }
])
```

# Questions

- Is there a way of accessing the `Response` object created by the `redirect` call from a loader/Element of the target route?
- If not: using loaders like this, is there another way to pass an error message to the route using `React Router`'s built-ins?
- If not: would it be worth opening a feature request for?

**Comment by kiliman** [comment]
Remix doesn't give you direct access to the response. If you need access to the redirect reason, then you can include it in the search params.

```ts
async function authenticatedPageLoader() {
  // Some auth check fails...
  return redirect("/login?reason=Please log in!");
} 

// This one receives `Params` and a `Request` without any information about the reason why the redirect happened which is passed to the `Response`.
async function loginLoader({params, request}) {
  // Explain the user that they need to be logged in to visit "/some/authenticated/page"
  const url = new URL(request.url)
  const reason = url.searchParams.get('reason') ?? 'Unknown reason'
  // ...
} 
```

**Reply by svenvandescheur** [reply]
This was a solution that came to mind, however this produces quite nasty URL's when the reason is longer.  Would love to have access to the redirected response as feature.

**Reply by brophdawg11** [reply]
I agree with @kiliman here in that we largely try to follow the browser `Request`/`Response` lifecycle you'd have in a normal MPA.  The subsequent request made by the redirect has no access to the response that initiated it.  

If a URL parameter is not suitable, the usual solution in that flow would be a [flash](https://remix.run/docs/en/main/utils/sessions#sessionflashkey-value) cookie.  The response sets a cookie that gets sent along with the subsequent redirect request and can be read - thus proxying information from the redirect response to the redirect request.

In a SPA, I would suggest doing something similar via localStorage or an in-memory cache.

**Reply by svenvandescheur** [reply]
I had another thought about this:

 > include it in the search params

Using the search params is not ideal as it is quite likely to introduce security issues (one could easily craft an url which shows a message asking the user to navigate to a hacker controlled website). Using a system of status codes and associated (known) messages can reduce this risk but reduces functionality. 

 >  localStorage or an in-memory cache

This could work but could easily introduce bugs, e.g. when certain data is not removed correctly or is overwritten by another reponse. Using this correctly would require the application to almost introduce something like an event queue in order to explain the user: "your session has expired" which feels to me as an over engineered solution (needlessly complex) for the job.

It's not unusual (in various domains) to have some kind of mechanism in place to get the original information about, for instance a request. We see such behaviour in reverse proxy's (e.g. [X-Forwarded-Host](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Host), [X-Forwarded-For](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For)) or even redirect information ([X-Redirect-By](https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/PageTypes/RedirectHeaders.html)) in frameworks. 

I think that providing access to the original response, one way or another is not something that would divert from the idea of maintaining a request/response like system (which I like btw) while making certain feature implementatinos a lot easier and possibly more secure.

**Reply by brophdawg11** [reply]
> This could work but could easily introduce bugs

The idea of a "flash" message has been around for a very long time.  No software is 100% bug free, but I'm confident you could achieve your use-case using a "flash" pattern.  Also, being that you are client-side and have a shared memory scope between your different loaders you could just as easily share the entire `Response` from the redirect and read it on the subsequent request.  I don't see any compelling reason for React Router to need to proxy `Response` instance through when this use case is solvable in user-land today.

**Reply by svenvandescheur** [reply]
> > This could work but could easily introduce bugs
> 
> The idea of a "flash" message has been around for a very long time. No software is 100% bug free, but I'm confident you could achieve your use-case using a "flash" pattern. Also, being that you are client-side and have a shared memory scope between your different loaders you could just as easily share the entire `Response` from the redirect and read it on the subsequent request. I

It's surely possible, however, as far as I know react-router doesn't set cookies, so some other way of communicating should be implemented, possibly localStorage (which might not be always available on all environments) or some sort of global, which is considered more ore less an anti pattern. A direct reference from react-router would IMO a lot more of a clean approach.

> don't see any compelling reason for React Router to need to proxy `Response` instance through when this use case is solvable in user-land today.

I guess that applies to a lot/everything what remix-run (or react-router) does. One could "easily" implement a router in userland without the need of a library. Having "batteries" includes providing the heavy lifting is a useful tool.

I thing the given use-case (along with others) isn't necessarily exotic and the having to resort to state management (one way or another) to bypass this restriction is... doable... but annoying...

At the moment I'm not aware of any pure, "clean" way, without the use of side-effects to get, let's say an error message. From a loader getting a 403 to the redirect target's loader.

**Reply by brophdawg11** [reply]
> Having "batteries" includes providing the heavy lifting is a useful tool

This is why we have Proposals!  If this proposal gets enough upvotes proving there is enough real world need for this feature, then it can be considered for implementation. Until then, it's doable in userland.

**Comment by svenvandescheur** [comment]
I changed the category from Q&A to proposals as the requested feature does not seem to be available. I hacked on a bit on the `@remix-run/router@1.17.0` tag and got this to work locally as a quick PoC. I'll see if I can make this work on the latest branch and possibly open a draft PR.

I'm new to the codebase so I probably missed a lot of scenario's though...

**Comment by svenvandescheur** [comment]
https://github.com/remix-run/react-router/pull/11722
